release/dist/qnap/pkgs.go

```diff
@@ -148,11 +110,103 @@ type qnapBuilds struct {
+func getQnapBuilds(b *dist.Build, signer *signer) *qnapBuilds {
+	return b.Extra(qnapBuildsMemoizeKey{}, func() any {
+		builds := new(qnapBuilds)
+		if err := builds.setUpTmpDir(b, signer); err != nil {
+			panic(fmt.Errorf("setUpTmpDir: %v", err))
+		}
+		return builds
+	}).(*qnapBuilds)
+}
+
+//go:embed all:files
+var buildFiles embed.FS
+
+// setUpTmpDir copies the contents of the buildFiles embedded FS into the local
+// m.tmpDir for building.
+//
+// We do this to allow for this tailscale.com/release/dist/qnap package to be
+// used from both the corp and OSS repos. When built from OSS source directly,
+// this is a superfluous extra step, but when imported as a go module to another
+// repo (such as corp), we must do this to allow for the module's build files
+// to be reachable and editable from docker.
+//
+// This runs only once per dist.Build instance, is shared by all qnap targets,
+// and gets cleaned up upon close of the dist.Build.
+//
+// When a signer is provided, setUpTmpDir also sets up the qpkg signature files
+// in qbuild's expected location within m.tmpDir.
+func (m *qnapBuilds) setUpTmpDir(b *dist.Build, signer *signer) error {
+	return b.Once(fmt.Sprintf("set-up-qnap-tmp-build-dir"), func() error {
+		log.Print("Setting up qnap tmp build directory")
+
+		m.tmpDir = filepath.Join(b.Repo, "tmp-qnap-build")
+		b.AddOnCloseFunc(func() error {
+			return os.RemoveAll(m.tmpDir)
+		})
+
+		if err := fs.WalkDir(buildFiles, "files", func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+			outPath := filepath.Join(m.tmpDir, path)
+			if d.IsDir() {
+				if err := os.MkdirAll(outPath, 0755); err != nil {
+					return err
+				}
+				return nil
+			}
+			file, err := fs.ReadFile(buildFiles, path)
+			if err != nil {
+				return err
+			}
+			if err := os.WriteFile(outPath, file, 0755); err != nil {
+				return err
+			}
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		if signer != nil {
+			log.Print("Setting up qnap signing files")
+
+			key, err := os.ReadFile(signer.privateKeyPath)
+			if err != nil {
+				return err
+			}
+			cert, err := os.ReadFile(signer.certificatePath)
+			if err != nil {
+				return err
+			}
+
+			// QNAP's qbuild command expects key and cert files to be in the root
+			// of the project directory (in our case release/dist/qnap/Tailscale).
+			// So here, we copy the key and cert over to the project folder for the
+			// duration of qnap package building and then delete them on close.
+
+			keyPath := filepath.Join(m.tmpDir, "files/Tailscale/private_key")
+			if err := os.WriteFile(keyPath, key, 0400); err != nil {
+				return err
+			}
+			certPath := filepath.Join(m.tmpDir, "files/Tailscale/certificate")
+			if err := os.WriteFile(certPath, cert, 0400); err != nil {
+				return err
+			}
+
+			b.AddOnCloseFunc(func() error {
+				return errors.Join(os.Remove(keyPath), os.Remove(certPath))
+			})
```

Comment:
since we will clean up `m.tmpDir` on close, this is no longer needed, right?

Reply:
Yeah exactly, these get added to m.tmpDir and we clean up that whole directory.